Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -857,33 +874,50 @@ void NetworkStorageManager::setItem(IPC::Connection& connection, StorageAreaIden
     ASSERT(!RunLoop::isMain());
 
     bool hasQuotaError = false;
-    if (auto storageArea = m_storageAreaRegistry->getStorageArea(identifier)) {
-        auto result = storageArea->setItem(connection.uniqueID(), implIdentifier, String { key }, WTFMove(value), WTFMove(urlString));
-        if (!result)
-            hasQuotaError = (result.error() == StorageError::QuotaExceeded);
-    }
+    auto storageArea = m_storageAreaRegistry->getStorageArea(identifier);
+    if (!storageArea)
+        return completionHandler(hasQuotaError);
 
+    auto result = storageArea->setItem(connection.uniqueID(), implIdentifier, String { key }, WTFMove(value), WTFMove(urlString));
+    if (!result)
+        hasQuotaError = (result.error() == StorageError::QuotaExceeded);
     completionHandler(hasQuotaError);
+
+    writeOriginToFileIf(storageArea->origin(), [&]() {
```

Comment:
Looks a little risky to call this *after* calling the completion handler? Are we sure the completion handler cannot cause any work that would require the origin to be written?

Reply:
This is a message handler, so the completion handler is just sending reply